include/seqan/stream/iostream_zip_impl.h

```diff
@@ -29,645 +13,530 @@
-
-        do {
-            if ( m_zip_stream.avail_in == 0 ) count = fill_input_buffer();
-
-            if ( m_zip_stream.avail_in ) {
-                m_err = inflate( &m_zip_stream, Z_SYNC_FLUSH );
+        m_err = deflate(&m_zip_stream, flush_mode);
+        if (m_err == Z_OK || m_err == Z_STREAM_END)
+        {
+            written_byte_size = static_cast<std::streamsize>(m_output_buffer.size()) - m_zip_stream.avail_out;
+            total_written_byte_size += written_byte_size;
+
+            // ouput buffer is full, dumping to ostream
+            m_ostream.write((const char_type *) &(m_output_buffer[0]),
+                            static_cast<std::streamsize>(written_byte_size / sizeof(char_type) * sizeof(byte_type)));
+
+            // checking if some bytes were not written.
+            if ((remainder = written_byte_size % sizeof(char_type)) != 0)
+            {
+                // copy to the beginning of the stream
+                std::memmove(&(m_output_buffer[0]),
+                             &(m_output_buffer[written_byte_size - remainder]),
+                             remainder);
             }
-        } while ( m_err == Z_OK && m_zip_stream.avail_out != 0 && count != 0 );
 
-        // updating crc
-        m_crc = crc32(
-            m_crc,
-            (byte_buffer_type)buffer_,
-            buffer_size_ - m_zip_stream.avail_out / sizeof( char_type )
-        );
-        std::streamsize n_read = buffer_size_ - m_zip_stream.avail_out / sizeof( char_type );
-
-        // check if it is the end
-        if ( m_err == Z_STREAM_END ) put_back_from_zip_stream();
-
-        return n_read;
+            m_zip_stream.avail_out = static_cast<uInt>(m_output_buffer.size() - remainder);
+            m_zip_stream.next_out = &m_output_buffer[remainder];
+        }
     }
-
-    template<
-        typename Elem,
-        typename Tr,
-        typename ElemA,
-        typename ByteT,
-        typename ByteAT
-    >
-    void basic_unzip_streambuf< Elem, Tr, ElemA, ByteT, ByteAT >::put_back_from_zip_stream()
+    while (m_err == Z_OK);
+
+    m_ostream.flush();
+
+    return total_written_byte_size;
+}
+
+template <typename Elem,
+          typename Tr,
+          typename ElemA,
+          typename ByteT,
+          typename ByteAT>
+std::streamsize basic_zip_streambuf<Elem, Tr, ElemA, ByteT, ByteAT>::flush()
+{
+    return flush(Z_SYNC_FLUSH);
+}
+
+template <typename Elem,
+          typename Tr,
+          typename ElemA,
+          typename ByteT,
+          typename ByteAT>
+std::streamsize basic_zip_streambuf<Elem, Tr, ElemA, ByteT, ByteAT>::flush_finalize()
+{
+    return flush(Z_FINISH);
+}
+
+template <typename Elem,
+          typename Tr,
+          typename ElemA,
+          typename ByteT,
+          typename ByteAT>
+basic_unzip_streambuf<Elem, Tr, ElemA, ByteT, ByteAT>::basic_unzip_streambuf(
+    istream_reference istream_,
+    size_t window_size_,
+    size_t read_buffer_size_,
+    size_t input_buffer_size_
+    ) :
```

Comment:
move `) :` to line above?

also, are the sizes const? If yes, can we mark them as such?